### ISSUE TITLE
Delete files are not removed after running Iceberg maintenance ops

### ISSUE THREAD

**Issue by szisiu** [trinodb/trino#24086]
Related to https://github.com/trinodb/trino/pull/23801;

After I've upgraded Trino from v459 to v464 I've noticed a big jump in number of files (delete files) in my cloud storage bucket (I use Iceberg connector).

My `MERE INTO` queries on partitioned table produce some position deletes (delete files). 

> Position deletes are local to a partition. OPTIMIZE supports only enforced predicates which select whole partitions. Therefore, we can clean up position deletes in OPTIMIZE when there are no path or file_modified_time predicates

Does it mean that when I run for example a series of maintenance ops like:

```sql
-- compact files that are less than 1gb and touched within the last 2 days (might not touch the whole partition!)
ALTER TABLE ${TBL} EXECUTE optimize(file_size_threshold => '1024MB') WHERE "$file_modified_time" >= date_trunc('day',current_date) - INTERVAL '2' DAY;
ALTER TABLE ${TBL} EXECUTE expire_snapshots(retention_threshold => '60s'); -- 60s set in trino config and in session
ALTER TABLE ${TBL} EXECUTE remove_orphan_files(retention_threshold => '60s');
```

delete files will NOT be pruned? Because after the Trino upgrade they are not.

> If you've already run OPTIMIZE before and the delete file is no longer referenced by any active data file, then the delete files are removed only when the minimum sequence number of data in the table becomes greater than the sequence number of the deletes.

Where I can inspect the `sequence number`? `${TBL}$metadata_log_entries`?

If yes, then I have something like this:

![image](https://github.com/user-attachments/assets/b328c896-1e21-4a99-aaf1-2de26405f011)

and in `${TBL}$files`:

![Image](https://github.com/user-attachments/assets/9db714c9-db41-490e-84ae-f319a63caa0c)

Recently, I've ran :

```sql
ALTER TABLE ${TBL} EXECUTE optimize(file_size_threshold => '1024MB');
ALTER TABLE ${TBL} EXECUTE expire_snapshots(retention_threshold => '60s');
ALTER TABLE ${TBL} EXECUTE remove_orphan_files(retention_threshold => '60s');
```

and delete files are still there.

Is there a better way to check if a delete file is no longer referenced by an active data file?

What I am missing? 

Let me know if you need more details.

**Comment by raunaqmorarka** [comment]
@szisiu can you try release 463 ?
I'm not sure that there are any adverse effects of https://github.com/trinodb/trino/pull/23801, but https://github.com/trinodb/trino/pull/23864 may be a factor

**Comment by szisiu** [comment]
@raunaqmorarka, I've tested trino v463 and it looks like you are right (docker-compose project with minio/hms/mysql).

steps to reproduce (on trino v463):

- create tables (target one with partitioning)

```sql
CREATE SCEMA iceberg.testschema 
    WITH (location = 's3a://testbucket/testschema');

CREATE TABLE IF NOT EXISTS "iceberg"."testschema"."sourcetbl" (
    "user" VARCHAR,
    "id" integer,
    "upserted" timestamp(6) with time zone
)
WITH (
    "format" = 'PARQUET',
    "format_version" = 2
);

CREATE TABLE IF NOT EXISTS "iceberg"."testschema"."targettbl" (
    "user" VARCHAR,
    "id" integer,
    "upserted" timestamp(6) with time zone
)
WITH (
    "format" = 'PARQUET',
    "format_version" = 2,
    "partitioning" = ARRAY['upserted']
);
```

-  fill the source table

```sql
INSERT INTO "iceberg"."testschema"."sourcetbl" VALUES
    ('u1', 1, TIMESTAMP '2024-11-12 14:39:20.228+01:00');
```

- fill the target table

```sql
MERGE INTO "iceberg"."testschema"."targettbl" "t"
USING (
    SELECT
        "max_by"("user", "upserted") "user",
        "max_by"("upserted", "upserted") "upserted",
        "max_by"("id", "upserted") "id"
    FROM "iceberg"."testschema"."sourcetbl"
    GROUP BY "id"
) "s"
ON "t"."id" = "s"."id"
WHEN MATCHED THEN UPDATE SET
    "user" = "s"."user",
    "id" = "s"."id",
    "upserted" = "s"."upserted"
WHEN NOT MATCHED THEN INSERT (
    "user",
    "id",
    "upserted"
)
VALUES (
    "s"."user", 
    "s"."id", 
    "s"."upserted"
)
```

![Image](https://github.com/user-attachments/assets/33fe003f-af24-4bea-bf93-675da175ea83)

- run MERGE INTO second time so delete files will be created

![Image](https://github.com/user-attachments/assets/c52f2627-666a-4122-a55e-a3afb0877ae3)

note the 1.2KB delete file

- when running optimize with file_modified_time predicate and other maintenance ops

```sql
ALTER TABLE "iceberg"."testschema"."targettbl" EXECUTE optimize(file_size_threshold => '1024MB') WHERE "$file_modified_time" >= date_trunc('day',current_date) - INTERVAL '1' DAY;
```

![Image](https://github.com/user-attachments/assets/20e616ac-8a76-456a-8927-ece0c3a1deff)

delete file is still there

```sql
SET session iceberg.expire_snapshots_min_retention = '60s';
ALTER TABLE "iceberg"."testschema"."targettbl" EXECUTE expire_snapshots(retention_threshold => '60s');

SET session iceberg.remove_orphan_files_min_retention = '60s';
ALTER TABLE "iceberg"."testschema"."targettbl" EXECUTE remove_orphan_files(retention_threshold => '60s');
```

![Image](https://github.com/user-attachments/assets/9fdb7d81-bc58-4725-baf1-508bfd1eb9c3)


- when running optimize (without predicates) and other maintenance ops

```sql
ALTER TABLE "iceberg"."testschema"."targettbl" EXECUTE optimize(file_size_threshold => '1024MB');

SET session iceberg.expire_snapshots_min_retention = '60s';
ALTER TABLE "iceberg"."testschema"."targettbl" EXECUTE expire_snapshots(retention_threshold => '60s');

SET session iceberg.remove_orphan_files_min_retention = '60s';
ALTER TABLE "iceberg"."testschema"."targettbl" EXECUTE remove_orphan_files(retention_threshold => '60s');
```

![Image](https://github.com/user-attachments/assets/3d5448a6-9e37-4780-a9a4-90e3a7b475fc)

So when using trino v463 all files are cleaned up and only single data file is there (for the example above). 

On trino v464 delete files are not cleaned up.

To sum up, it looks like some changes introduced in trino v464 have a bug (https://github.com/trinodb/trino/pull/23864?).

**Comment by hashhar** [comment]
FYI @findinpath

**Comment by raunaqmorarka** [comment]
I think this is working as intended. I checked that we clean up deletes as expected when `file_modified_time` filter is not used.
When `file_modified_time` or `path` filter is used, we don't clean up deletes because it's possible for delete files to be attached to multiple files within a partition and this filter may select only a subset of data files in the partition (this behaviour is unchanged from previous releases).
Although we're not cleaning up the delete file, the new data file after optimize has the deletes applied and so we don't incur perf penalty of applying deletes on read.
When we skip removing the delete file, it will never get removed until the min sequence number of all active data files in the table becomes greater than the sequence number of that delete file (`org.apache.iceberg.MergingSnapshotProducer#apply`). Essentially all the data files that were present when the delete file was originally written, need to get rewritten, so that MergingSnapshotProducer can "prove" that no data file could refer to this delete anymore and remove it.
In the past releases, you're not hitting this issue because of a side effect of not having the optimization from https://github.com/trinodb/trino/pull/23864
In previous releases after the first OPTIMIZE call with `file_modified_time` filter will not clean up the delete and have the same end state as current release. However, running OPTIMIZE a second time will result in another rewrite of the data file which will allow clean up of the delete. This rewrite of the data file was actually unnecessary as it's a single file without a delete and doesn't benefit from rewrite. But it has the side effect of moving the lowest sequence_number of data files in the table beyond the sequence_number of the delete, which allows iceberg library to clean up the delete file. If you add another small file to your table and then re-run OPTIMIZE, you should see a clean up of the delete in current release as well.
I can even simulate the same issue in previous releases for an OPTIMIZE call on a table with multiple data files with a delete by tweaking `file_size_threshold` high enough for it to miss a subset (but not all) of the data files in the table.
So the 464 changes are not creating a new problem, but perhaps making it more likely that we will leave behind these "dangling" deletes.
My suggestion would be to drop `file_modified_time` filter. That will avoid hitting this scenario. I don't see any benefit of including that filter because we scan all the iceberg metadata regardless. The logic around file size, detection of deletes and awareness about single file per partition is sufficient to ensure that we are not rewriting any files needlessly.

**Comment by szisiu** [comment]
Thanks @raunaqmorarka for the detailed response.

> When we skip removing the delete file, it will never get removed until the min sequence number of all active data files in the table becomes greater than the sequence number of that delete file (org.apache.iceberg.MergingSnapshotProducer#apply).

I need to dig into that code. I was not able to remove the deleted files even if more data were written to the table (between optimize calls).

**Comment by raunaqmorarka** [comment]
You can also look into running https://iceberg.apache.org/docs/latest/spark-procedures/#rewrite_data_files with `remove-dangling-deletes` enabled, that should in theory clean up any dangling deletes. Though my main suggestion would be to just drop `file_modified_time` filter from optimize.

**Comment by szisiu** [comment]
Unfoutunately I do not use Spark.

For some time now I am using trino v466 and removed file_modified_time filter from optimize - it looks good.

Closing.

**Comment by vburenin** [comment]
This is not resolved and the issue persist for me in Trino 470.

**Comment by dvaldivia** [comment]
+1

**Comment by sharon-clue** [comment]
+1

**Comment by Nazarii** [comment]
It occurs for me in Trino 479 as well, I need to do a separate automated cleanup to drop zombie files.

**Comment by fcfangcc** [comment]
It occurs for me in Trino 483 as well,  I don`t use "$file_modified_time" this problem still exists.  Just
```
ALTER TABLE test_table EXECUTE optimize(file_size_threshold => '128MB')
```
